### ISSUE TITLE
Reproduction gap vs released baseline (Qwen3VL-OFT-Robotwin2-All / ckpt-140000)

### ISSUE THREAD

**Issue by zxduan-r** [starVLA/starVLA#190]
## Overview

We verified the released baseline weights and attempted to reproduce the training. This issue summarizes our findings and asks for help narrowing down the performance gap.

---

## 1. Verification of Released Weights

We ran evaluation of the released `Qwen3VL-OFT-Robotwin2-All / ckpt-140000` weights on RoboTwin 2.0 using the same evaluation protocol as the README (100 episodes per task). Results are largely consistent with the README:

| Task | README (Easy) | Our Eval (clean) | README (Hard) | Our Eval (rand) |
|---|---|---|---|---|
| grab_roller | 100% | 100% | 100% | 100% |
| handover_mic | 98% | 97% | 96% | 96% |
| lift_pot | 100% | 100% | 100% | 100% |
| move_can_pot | 91% | 94% | 90% | 94% |
| open_laptop | 98% | 97% | 100% | 100% |
| pick_dual_bottles | 91% | 88% | 93% | 95% |
| place_dual_shoes | 91% | 91% | 89% | 95% |
| place_object_basket | 93% | 90% | 94% | 91% |
| place_phone_stand | 86% | 86% | 95% | 84% |
| put_bottles_dustbin | 90% | 88% | 85% | 91% |
| put_object_cabinet | 89% | 90% | 91% | 94% |
| stack_blocks_two | 100% | 100% | 100% | 99% |
| stack_bowls_two | 99% | 99% | 100% | 99% |

The released weights reproduce well. ✅

---

### a. Dataset alignment

We compared `dataset_statistics.json` from both checkpoints:

| Metric | Released | Ours |
|---|---|---|
| `num_trajectories` | 27,500 | 27,500 |
| `num_transitions` | 6,075,103 | 6,102,603 |
| `data_mix` | `robotwin_all_50` | `robotwin_all_50` |

Trajectories are identical. The +27,500 transitions difference (exactly 1 per trajectory) is likely a minor version difference in the episode `length` field of `episodes.jsonl`. Action statistics (`q01`/`q99`) are very close.

### b. Training configuration

Both runs use essentially the same config:

| Parameter | Released | Ours |
|---|---|---|
| `framework` | QwenOFT | QwenOFT |
| `base_vlm` | Qwen3-VL-4B-Instruct | Qwen3-VL-4B-Instruct |
| `action_model_type` | DiT-B | DiT-B |
| `action_hidden_dim` | 2560 | 2560 |
| `future_action_window_size` | 49 | 49 |
| `per_device_batch_size` | 4 | 4 |
| `num_processes` | 8 | 8 |
| `learning_rate (base)` | 1e-5 | 1e-5 |
| `learning_rate (action_model)` | 1e-4 | 1e-4 |
| `max_train_steps` | 150,000 | 150,000 |
| `num_warmup_steps` | 5,000 | 5,000 |
| `lr_scheduler` | cosine_with_min_lr | cosine_with_min_lr |
| `attn_implementation` | **sdpa** | **eager** |
| `freeze_modules` | `true` (no effect*) | `null` |

> \* `freeze_modules: true` (boolean) is ignored by the code which checks `type(freeze_modules) == str`, so both runs trained with all parameters unfrozen.

The only notable difference is `attn_implementation: sdpa` vs `eager`.

### c. Reproduction results (ckpt-140000, 11/13 tasks completed so far)

| Task | Released (clean) | Ours (clean) | Released (rand) | Ours (rand) |
|---|---|---|---|---|
| grab_roller | 100% | 100% | 100% | 99% |
| handover_mic | 97% | 97% | 96% | 100% |
| lift_pot | 100% | 99% | 100% | 97% |
| move_can_pot | 94% | 69% | 94% | 69% |
| open_laptop | 97% | 91% | 100% | 97% |
| pick_dual_bottles | 88% | 46% | 95% | 58% |
| place_dual_shoes | 91% | 41% | 95% | (running) |
| place_object_basket | 90% | 77% | 91% | 78% |
| place_phone_stand | 86% | 71% | 84% | 81% |
| stack_blocks_two | 100% | 19% | 99% | 24% |
| stack_bowls_two | 99% | 94% | 99% | 92% |
| **Avg (11 tasks)** | **94.7%** | **73.1%** | **95.8%** | **79.5%** |

There is a ~20% gap on average, with particularly large drops on `stack_blocks_two` (-81%), `pick_dual_bottles` (-42%), `place_dual_shoes` (-50%), and `move_can_pot` (-25%).

---

## 3. Questions

We're not sure what's causing this gap given the configs are nearly identical. Could you help with:

1. Was `attn_implementation: sdpa` important? Could switching from `eager` explain such a large difference?
2. Was there any data preprocessing step (e.g. re-generating `episodes.jsonl` with different `length` values) that we might have missed?
3. Are there any other known factors (e.g. specific DeepSpeed config, ZeRO stage, gradient checkpointing settings) that are critical for reproduction?
4. Is there a loss curve available for the released run that we could compare against?

Thanks in advance!

**Comment by Axi404** [comment]
1. There is no evidence that this setting is important, and we also lack ablation studies.
2. Probably not.
3. Probably not.
4. Yes. See: https://api.wandb.ai/links/axi-the-cat/7fus3mhz

However, what might really matter is the batch size (or other training settings). We trained this model on 6×8 A800 GPUs with a batch size of 4 per GPU. What setup are you using?

If you trained your model for less than our batch size, 150k steps might not able to ran 1 epoch

**Comment by zxduan-r** [comment]
Thanks for the detailed reply and the loss curve link!

Our setup is also 8 GPUs with batch size 4 per GPU, so the total batch size should be identical to yours.

**Comment by Axi404** [comment]
Just to clarify, our setup uses **48 GPUs (6 nodes × 8 A800 GPUs)**. So compared to your **8-GPU setup**, your effective epoch progress would be **about 6× smaller** under the same step count. In other words, your run likely corresponds to **less than ~0.2 epoch** on our scale.

**Comment by Nemo-1024** [comment]
I hope the StarVLA project team can clarify how many GPUs were used for each benchmark and report the total batch size, rather than only providing the per-device batch size.

**Comment by JinhuiYE** [comment]
<img width="1130" height="302" alt="Image" src="https://github.com/user-attachments/assets/7dad32a2-5818-4bb9-9911-c6380f7844ed" />

Sure!

**Comment by zxduan-r** [comment]
感谢作者的回复和 loss curve！这正好解答了我们的疑惑。这里补充一下我们后续的实验结果，希望对其他遇到同样问题的用户有参考价值。

---

## 评测说明

为了加快验证速度，我们后续采用了 **subset 评测**的方式（共 13 个 task，每个 task 10 episodes），而非官方的 100 episodes。小样本存在一定方差，但趋势基本一致。

**官方权重（ckpt-140000）subset 验证结果：**

| Task | clean (我们, 10ep) | rand (我们, 10ep) |
|---|---|---|
| grab_roller | 100% | 100% |
| handover_mic | 100% | 90% |
| lift_pot | 100% | 100% |
| move_can_pot | 90% | 70% |
| open_laptop | 100% | 100% |
| pick_dual_bottles | 70% | 90% |
| place_dual_shoes | 70% | 100% |
| place_object_basket | 90% | 90% |
| place_phone_stand | 90% | 90% |
| put_bottles_dustbin | 90% | 90% |
| put_object_cabinet | 100% | 100% |
| stack_blocks_two | 100% | 100% |
| stack_bowls_two | 100% | 80% |
| **Average** | **92%** | **92%** |

结果与官方 100 episodes 高度一致，证明 10 episodes subset 评测方式可用于快速对比。

---

## 我们的 16 卡复现尝试（global BS = 192，训练 150k steps）

按照作者的说明，我们使用 **16 卡（2 节点 × 8 GPU，per_device_bs=12）** 以使 global batch size 与官方对齐（192 = 12 × 16），训练 150k steps，评测 ckpt-140000：

| Task | clean | rand |
|---|---|---|
| stack_blocks_two | 10% | 0% |
| pick_dual_bottles | 50% | 40% |
| place_dual_shoes | 40% | 50% |
| move_can_pot | 70% | 70% |
| **Average (13 tasks)** | **~78%** | **~84%** |

与官方权重相比仍有约 **10~15% 的差距**，精细操作类任务（`stack_blocks_two`、`pick_dual_bottles`、`place_dual_shoes`）差距尤为明显。

---

## step_stride 实验：加速训练的意外收获

为了在有限 GPU 资源下加速训练，我们实验了 `step_stride`（采样间隔）超参数，意外发现效果有明显提升：

| 实验 | GPU | per_device_bs | global BS | step_stride | max_steps | ckpt | clean | rand |
|---|---|---|---|---|---|---|---|---|
| 1node_stride3 | 8 卡 | 24 | 192 | 3 | 50k | **40k** | **89%** | **90%** |
| 1node_stride5 | 8 卡 | 24 | 192 | 5 | 30k | 30k | 75% | 79% |

`stride=3` 在 ckpt-40k 时达到 89%/90%，是我们目前最佳的自训练结果，接近官方 48 卡权重（93%）。

---

希望这些结果对其他资源受限的用户有参考价值。如有其他发现会继续更新。

**Comment by Bariona** [comment]
Hi @zxduan-r , just curious what `step_stride` means here. And how much GPU mem is needed for 24 bs/gpu and also the training time?

**Comment by zxduan-r** [comment]
> Hi [@zxduan-r](https://github.com/zxduan-r) , just curious what `step_stride` means here. And how much GPU mem is needed for 24 bs/gpu and also the training time?

@Bariona 我们对数据集进行了间隔采样(step_stride)，这样相同epoch下，可以减少训练的steps； 我们是140G的gpu，单卡batch size可以到24

**Comment by zxduan-r** [comment]
@JinhuiYE @Nemo-1024 @Bariona @Axi404  
## 更新：全 50 task 评测结果 + 异常任务排查

### 我们复现权重的全 50 task 评测（2node_bs12, ckpt-140000）

配置：16 卡（2 节点 × 8 GPU），per_device_bs=12，global BS=192，max_steps=150k， 有一些task没跑完，但是对最终结果影响不大。

| | demo_clean | demo_randomized |
|---|---|---|
| **Total (50 tasks)** | **3834/4924 (78%)** | **3843/4723 (81%)** |

---

### 三个异常 task（准确率接近 0%）

我们发现三个 task，复现权重的准确率接近 0%，而官方权重表现良好：

| task | 我们 clean | 官方 clean | 我们 rand | 官方 rand |
|---|---|---|---|---|
| `blocks_ranking_rgb` | 0% | 99% | 0% | 98% |
| `stack_blocks_three` | 0% | 94% | 0% | 95% |
| `stack_blocks_two` | 1% | 100% | 2% | 99% |

我们用官方权重（`Qwen3VL-OFT-Robotwin2-All / ckpt-140000`）在这三个 task 上跑了 100 episodes，结果与 README 一致（99%/94%/100%），因此评测环境没有问题，异常是复现权重特有的。

这三个 task 的训练数据完整（Clean 50 ep，Randomized 500 ep），与其他 task 相同。

请问是否有什么特别的原因导致这三个 task 特别难复现？ 能给出一些排查方向吗？

**Comment by Nemo-1024** [comment]
我在我自己的训练日志中检查了blocks_ranking_rgb clean的情况，我发现我自己的模型在这个任务上，有些检查点是100%，而另一次不同参数的训练是2%，为此这可能是调参的问题，（但这确实很奇怪）。但我没有尝试过测试starVLA的权重，如果你当时保存了视频，或许可以分享下发生了什么。

**Comment by brightest66** [comment]
In the discussion, you referred to changing the batch size to match the author's configuration.

To maintain the same training dynamics and effective batch size, could you please clarify whether the `gradient_accumulation_steps` was also modified in tandem with the batch size change?

**Comment by JinhuiYE** [comment]
No gradient accumulation was used in the released run. The global batch size was achieved purely via data parallelism across 48 GPUs (6 nodes × 8 A800 GPUs, per_device_bs=4), as shown in the config screenshot above. If you're using fewer GPUs and want to match the effective batch, you can use gradient_accumulation_steps — but note that with DeepSpeed ZeRO, gradient accumulation and pure data-parallel batch scaling may not be fully equivalent in training dynamics. @zxduan-r's experiments in this thread (using stride sampling to compensate) may give useful reference points.

**Comment by liujiahao2077** [comment]
Regarding the experimental results of RoboTwin, was the VLM data provided in starVLA used during the training process?

**Comment by JinhuiYE** [comment]
no, only data from RoboTwin

**Comment by liujiahao2077** [comment]
thank you！
